src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/HeaderModelBinder.cs

```diff
@@ -45,26 +50,112 @@ public HeaderModelBinder(ILoggerFactory loggerFactory)
             _logger = loggerFactory.CreateLogger<HeaderModelBinder>();
         }
 
+        /// <summary>
+        /// Initializes a new instance of <see cref="HeaderModelBinder"/>.
+        /// </summary>
+        /// <param name="innerModelBinder">The <see cref="IModelBinder"/> which does the actual
+        /// binding of values.</param>
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
+        public HeaderModelBinder(ILoggerFactory loggerFactory, IModelBinder innerModelBinder)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            if (innerModelBinder == null)
+            {
+                throw new ArgumentNullException(nameof(innerModelBinder));
+            }
+
+            InnerModelBinder = innerModelBinder;
+            _logger = loggerFactory.CreateLogger<HeaderModelBinder>();
+        }
+
+        // to enable unit testing
+        internal IModelBinder InnerModelBinder { get; }
+
         /// <inheritdoc />
-        public Task BindModelAsync(ModelBindingContext bindingContext)
+        public async Task BindModelAsync(ModelBindingContext bindingContext)
```

Comment:
No reason for an `async` state machine when doing the old thing. Suggest making this method non-`async`. Either call `BindWithoutInnerBinder(...)` and return `Task.CompletedTask` or a new method (that is `async`).

But, don't worry about it if others thing the "old thing" is an extreme corner case.